DynamicTablesPkg/Include/X64NameSpaceObjects.h

```diff
@@ -223,9 +224,34 @@ typedef struct CmX64MadtInfo {
   ID: EX64ObjLocalApicX2ApicInfo
 */
 typedef struct CmX64LocalApicX2ApicInfo {
-  UINT32    ApicId;
-  UINT32    Flags;
-  UINT32    AcpiProcessorUid;
+  UINT32             ApicId;
+  UINT32             Flags;
+  UINT32             AcpiProcessorUid;
+
+  /** Optional field: Reference Token for the Cst info of this processor.
+      i.e. a token referencing a CM_ARCH_COMMON_CST_INFO object.
+  */
+  CM_OBJECT_TOKEN    CstToken;

```

Comment:
I think these should be token identifying a `CM_ARCH_COMMON_OBJ_REF` structure instead.

Indeed, it is likely that most of the CPUs on the platform will have the same set of C-States or P-States (or other). So instead of having all functions returning the same set of values for each CPU, it would be better to factorize these values in one object. For instance, all CPUs on the platform could implement the `_CST` method as:

```
Scope (\_SB) {
	Name (CST0, Package() {
		0, // Count
		Package() {
			Register,
			Type,
			Latency,
			Power
		},
	} // Name CST0
}

// and for each CPU:
Scope(\_PR.CPU0) {
  Method(_CST,0) {
	  Return(\_SB.CST0)
  }
}

Scope(\_PR.CPU1) {
  Method(_CST,0) {
	  Return(\_SB.CST0)
  }
}

Scope(\_PR.CPU2) {
  Method(_CST,0) {
	  Return(\_SB.CST0)
  }
}
...
```

This seems to be also how platform vendors have implemented their `_PSS`, `_PCT` methods.

Please check the `GenerateLpiStates()` and `CreateAmlLpiMethod()` functions

Reply:
Didnt opted the factorization for below reason.
1) to generate the code similar to what ACPI spec says and close to the acpi spec examples.
2) There are possibilities(hypothetical scenario) that each processor might support different set of C-State and P-State.
   Also different processor might have different values(in package) to put the processor into different C-State.

Reply:
Using a `CM_ARCH_COMMON_OBJ_REF` would still allow CPUs to have different C-States/P-states. However they would have to be written only once. Each CPU can then aggregate these C/P-states how they desire.

Please have a look at the Lpi-States configuration at [1]. The Juno platform has an asymmetric topology, so for most platforms Lpi-States configuration would be simpler to configure (and so would it be for the C/P-States).

[1] https://github.com/tianocore/edk2-platforms/blob/master/Platform/ARM/JunoPkg/ConfigurationManager/ConfigurationManagerDxe/ConfigurationManager.c

Reply:
I reviewed the code for the LPI state implementation and its categorization of LPI states for each CPU.
I believe the current implementation offers the following advantages over the factorization approach:

1. Simplicity:
   The current implementation generates the C-state code according to the ACPI specification, making it easy to understand and correlate with the spec.
   There are no additional named packages or namespaces defined beyond what the ACPI specification allows. While it may often generate similar CST states for all CPUs, it leverages the DynamicTablesPkg to generate the code during boot time rather than requiring developers to write it manually.

2. Avoidance of Namespace Conflicts:
   LPI factorization uses Lxxx (L000 to Lxxx) for each CPU.
   We cannot use the name Cxxx (C000 to Cxxx) as it is reserved for defining CPU device objects, which would create a namespace conflict.
   Using names like CST0/CS00 to CSTx/CS00 would not support the maximum processor limit or MAX_NODE_COUNT.
   Alternatively, using other letters such as Txxx/Wxx introduces new named packages and could potentially conflict with current or future ACPI specifications if similar names are introduced.

Reply:
Hello Abdul,

1.
I agree that this is a big plus is for developpers to describe C-States (and others) values with your structure. Whith what I suggest, it is more complex to describe the states and how the relate to each others.
Using intermediate names is not really an issue I think, as long as we return the correct values.
About the fact that most CPUs will return the same values, I think this weights more in favor of factorizing the asl code: after the generation, it allows to see more clearly which CPUs have the same set of values.

2.
I think the issue could easily be solved by creating a specific namespace for each object type. For instance, all _CST structures could be stored under a 'CST' scope. In this scope, all objects would have a unique name:
```
Scope (_SB) {
  Scope (CST) {
    Name (CST0, Package() {
      0, // Count
      Package() {
        Register,
        Type,
        Latency,
        Power
      },
    } // Name CST0

  Name (CST1, Package() {
      0, // Count
      Package() {
        Register,
        Type,
        Latency,
        Power
      },
    } // Name CST1

  ...

  } // Scope CST
}
```

Another concern that I have is that for arm, the DynamicTablesPkg is really useful to create ACPI tables for virtual machines. The information comes from a device tree.
For instance to describe idle states:
https://github.com/torvalds/linux/blob/f44d154d6e3d633d4c49a5d6aed8a0e4684ae25e/arch/arm64/boot/dts/mediatek/mt8183.dtsi#L511
```
idle-states {
	CPU_SLEEP: cpu-sleep {
		entry-latency-us = <200>;
		exit-latency-us = <200>;
		min-residency-us = <800>;
	};
	CLUSTER_SLEEP0: cluster-sleep-0 {
		entry-latency-us = <250>;
		exit-latency-us = <400>;
		min-residency-us = <1000>;
	};
	CLUSTER_SLEEP1: cluster-sleep-1 {
		entry-latency-us = <250>;
		exit-latency-us = <400>;
		min-residency-us = <1300>;
	};
};

cpu0: cpu@0 {
	cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP0>;
};
cpu1: cpu@1 {
	cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP0>;
};
cpu4: cpu@100 {
	cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP1>;
};
cpu5: cpu@101 {
	cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP1>;
};
```

If the DynamicTablesPkg framework had to parse a device tree to create an ACPI description of the idle states, the steps would be:
1. Parse the device tree to create CM_XXX objects
2. Do some modification if necessary, i.e. some information might not be available in the DT and we might need to complete it. Or we might also need to complete the CM_XXX objects with information coming from another source.
3. Generate an ACPI representation of the idle states from the CM_XXX objects.

With the version you suggest, we would store an aggregation of the different idle states for each CPU:
// For CPU0
AML_CST_INFO cpu0_idle_states[2] {
  CPU_SLEEP,
  CLUSTER_SLEEP0
}
// For CPU1
AML_CST_INFO cpu1_idle_states[2] {
  CPU_SLEEP,
  CLUSTER_SLEEP0
}

If during the phase 2, we need to modify/complete the information of CLUSTER_SLEEP0, this will be difficult to find back the structures to modify. Indeed, the dependency relation was lost during the parsing phase.

Using tokens allows to conserve the dependency relation until the last minute for the generation step. Also, there is no duplication of information. This makes step 2 easier.

Now I understand that amd might not use a device tree representation. However:
- it is likely that for the same use case (ACPI generation for virtual machines), the information might be given in the same format as the device tree does. So what I described for arm might happen for amd
- _CST (and other objects) are also used by other architectures, so if another architecture comes with this exact use case, the CM_XXX objects will have to be modified and this will break amd platforms